README
TorrentFlux is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

TorrentFlux is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TorrentFlux; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Visit TorrentFlux.com

View the INSTALL file for installation instructions.

CHANGELOG
TorrentFlux: PHP Torrent Download Manager
=================================================================
by TorrentFlux.com
=================================================================
Change Log

TorrentFlux v2.4
------------------
* Powered by BitTornado 0.3.18 to support encryption
* Removed TorrentSpy search plugin
* Cryptography functionality & enabled by default (by imrook)
* Added checks to prevent normal users from seeing sensitive information (by deadeyes)
* Updated MiniNova search plugin (by alx)
* Removed CVS files for BitTornado
* Converted \r\n to \n in all source files
* Bug fix for torrents dying after they're started (tornado shell args not escaped)
* Postgres 8 sql install scripts
* Updated MySQL install scripts (TYPE clause depreciated for storage engine declarations)
* Fixed bug preventing torrents from starting in simple mode

Bug fixes from Debian team
* Recursive call error in SearchEngineBase
* Python filename dependency
* Parameter check in directory browser to prevent traversal above download directory
* Update settings fails with blank search filter

TorrentFlux v2.3
------------------
* Fixed more Arbitrary File Creation/Overwrite/Deletion & Command 
  Execution Vulnerablity (by TF Users).
* Security fixes on htmlentities.
* Added security check for JPG support with GD libs to prevent 
  user lock out if JPG support is not included in GD libs.
  (Note GD support is not required and is only for graphic 
  Security Code login).
* Upgraded to BitTornado 3.17 (BitTornado 3.15 source is still 
  included in tar.gz if you want to change back to 3.15).
  

TorrentFlux v2.2
------------------
* Added optional Security Code for login security.
* Fixed Arbitrary File Creation/Overwrite/Deletion & Command 
  Execution Vulnerablity (by TF Users).
* Fixed Qmanager looping issue.
* Added multi select delete to the directory listing.
* Updated RSSReader to not show error when no RSS feeds are used.
* Updated Search Engine Modules and bug fixes.
* Updated maketorrent to allow changing of chunk size.
* Added security fixes for htmlentities in User-Agents and 
  request variables.
* Fixed torrent start when advanced torrent start is turned off.
* Added page refresh countdown.
* Fixed jump links on RSS feed page.
* Updated Admin Links to allow editing and sorting.
* Added Turkish and Russian language files.


TorrentFlux v2.1
----------------
* Addes support for reading file size of files over 2.0 GB.
* Added Queue Manager functionality
  this will allow an admin to specify the following.
       maximum torrent threads per server
       maximum torrent threads per user
       polling interval 
           amount of time the q manager waits in (seconds) 
           between each sweep of the queue
       enable / disable of queuing
  admins will have the ability to override the queuing
  when starting a torrent in the start dialog.
* Settings are now stored in a database table and can be 
  modified from the admin screens.
* Torrents will not be able to be deleted until they have 
  shutdown -- prevents hanging threads.
* Admin can specify what categories will be filtered out on
  torrent searches.
* User may continue browsing when downloading via package
  downloading.
* Using ADODB 4.80 for database connections.
* Using BitTornado 3.15 Linux Source.
* Support for Cookies for each user under profiles.
* Updated RSS feed and Fetching of Torrents over a URL.
* Added Rerequest Interval to the Startpop and Settings
* Added modular search engine.  Support for several 
  popular torrent search engines -- you can build your
  own.
* Search Engine settings and filtering.
* File Priority selection 
* Added btmakemetafile.py for Making Torrents
* Users may download the torrent file from the list.
* NFO / Text viewer
* Download torrent meta file from main torrent list.

TorrentFlux v2.0
-----------------
* Added Size to the main page to view the download size of the 
  torrent.
* Added natural sort order for directory and file listings.
* Fixed deleting so that file names with .. in it will delete.
* Download files via the directory listing.
* Tar/Zip and download directories on with a single click.
* Added max download rate as an option.
* Added Super Seeder option for advanced start.
* Added mouse over torrent status.
* Added Database Abstraction layer (using ADODB)
* Updated to use BitTornado-0.3.10 source.
* Added new torrent search.
* Dynamic URLs now download and use proper torrent name.
* Added Chinese, Tiawanese, Finnish, Galacian, and Estonian 
  language files.
* Option to show server load on main page.
* Added torrent health icons for torrents that are running.
* Each user now sees his torrents grouped together in their own 
  list.
* Dropped BitTorrent client from install package.  Includes 
  BitTornado only now.
  

TorrentFlux v1.5
-----------------
* Created login screen -- no longer uses Basic Authentication 
  through the headers.  PHP no longer has to run as a module in 
  Apache.
* Added loggoff button.
* Storage bar to display free drive space.
* Download details now shows distributed seeds (BitTornado only).
* Download details now show download and upload amounts.
* Updated the torrent grabing via URL -- uses curl or fopen
  depending on what is available to server.  Much faster.
* Endless seeding (0% share kill) bug has been fixed.
* Minor bug fixes.
* Added Polish language file.
* Changed how torrents are stopped.  No longer uses kill command 
  but now messages the torrent client to shutdown -- stats are 
  now reported to the tracker when torrent is stopped.
* Enhanced directory browsing with file type icons.


TorrentFlux v1.4
-----------------
* The RSS Feed page now has jump links to the different sections 
  of the page.
* Better directory browsing.
* More complete logging for Admins.
* Total Upload and Download rates displayed at bottom of page.
* Replaced time stamp column with Estimated Time.
* Users may specify Max Upload Rate when running torrent.

html/multi.php
<?php

/*************************************************************
*  TorrentFlux - PHP Torrent Manager
*  www.torrentflux.com
**************************************************************/
/*
   This file is part of TorrentFlux.

   TorrentFlux is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   TorrentFlux is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with TorrentFlux; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

include_once("config.php");
include_once("functions.php");

# what to do?
$action = getRequestVar("action");
if($action == "torrent" || $action == "data")
{
   $torrent = getRequestVar("torrent");
   foreach($torrent as $key => $element)
   {
      # since we only have the .torrent file, get the stat
      $alias = getAliasName($element).".stat";
      # this is acctualy a copy and paste from index.php since its not a function
      # and we need to call it serveral times in a row
      $element = urldecode($element);

      if (($cfg["user"] == getOwner($element)) || IsAdmin())
      {
           # the user is the owner of the torrent -> delete it
           # FIRST delete data, then the torrent
           if ($action == "data")
           {

               require_once('BDecode.php');
               $ftorrent=$cfg["torrent_file_path"].$element;
               $fd = fopen($ftorrent, "rd");
               $alltorrent = fread($fd, filesize($ftorrent));
               $btmeta = BDecode($alltorrent);
               $delete = $btmeta['info']['name'];
               if(trim($delete) != "")
               {
                   $delete = $cfg['user']."/".$delete;

                   # this is accutaly from dir.php - its not a function, and we need to call it serval times

                   $del = stripslashes(stripslashes($delete));

                   if (!ereg("(\.\.\/)", $del))
                   {

                       avddelete($cfg["path"].$del);

                       $arTemp = explode("/", $del);
                       if (count($arTemp) > 1)
                       {
                           array_pop($arTemp);
                           $current = implode("/", $arTemp);
                       }
                       AuditAction($cfg["constants"]["fm_delete"], $del);
                   }
                   else
                   {
                       AuditAction($cfg["constants"]["error"], "ILLEGAL DELETE: ".$cfg['user']." tried to delete ".$del);
                   }
               }
           }
           @unlink($cfg["torrent_file_path"].$element);
           @unlink($cfg["torrent_file_path"].$alias);
           @unlink($cfg["torrent_file_path"].getAliasName($element).".prio");
           AuditAction($cfg["constants"]["delete_torrent"], $element);

       }
       else
       {
           AuditAction($cfg["constants"]["error"], $cfg["user"]." attempted to delete ".$element);
       }
   }
}
else if ( $action == "fileDelete" )
{
    $file = getRequestVar("file");
    // Lets delete some files
    if(is_array($file))
    {
        foreach($file as $key => $element)
        {
            $element = urldecode($element);
            delFile($element);
        }
    }
}

if ( isset($_SERVER["HTTP_REFERER"]) )
{
    header("location: ".$_SERVER["HTTP_REFERER"]);
}
else
{
    header("location: index.php");
}


function delFile($del)
{
    global $cfg;

    if(IsAdmin($cfg["user"]) || preg_match("/^" . $cfg["user"] . "/",$del))
    {
        // Yes, then delete it

        // we need to strip slashes twice in some circumstances
        // Ex.  If we are trying to delete test/tester's file/test.txt
        //    $del will be "test/tester\\\'s file/test.txt"
        //    one strip will give us "test/tester\'s file/test.txt
        //    the second strip will give us the correct
        //        "test/tester's file/test.txt"

        $del = stripslashes(stripslashes($del));

        if (!ereg("(\.\.\/)", $del))
        {
            avddelete($cfg["path"].$del);

            $arTemp = explode("/", $del);
            if (count($arTemp) > 1)
            {
                array_pop($arTemp);
                $current = implode("/", $arTemp);
            }
            AuditAction($cfg["constants"]["fm_delete"], $del);
        }
        else
        {
            AuditAction($cfg["constants"]["error"], "ILLEGAL DELETE: ".$cfg['user']." tried to delete ".$del);
        }
    }
    else
    {
        AuditAction($cfg["constants"]["error"], "ILLEGAL DELETE: ".$cfg['user']." tried to delete ".$del);
    }
}

?>

